### DISCUSSION TITLE
Improve unit conversion for sensor entities

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#534] [Entity Models] [answered]
Proposal:
Move temperature unit conversion from `Entity._async_write_ha_state` to sensor base class `SensorEntity`.

Suggested implementation alternatives:
- Add new properties `converted_state` and `converted_unit_of_measurement` to `Entity` with default implementations which simply returns `self.state`, `self.unit_of_measurement`, override those in `SensorEntity` to do unit conversion based on device class and update `Entity._async_write_ha_state` to read from `self.converted_state` instead of `self.state` and from `converted_unit_of_measurement` instead of `unit_of_measurement`.
**Pro:**
  - Very simple to implement

  **Con:**
  - `Entity` has more sensor specific support, not less, after this change

- Make `SensorEntity.state` property handle unit conversions and update all sensors to no longer override the `state` and `unit_of_measurement` properties, but instead override two new properties named `sensor_value` and `sensor_unit_of_measurement`, or similar, which represent the values read from the sensor.
**Pro:** 
   - No `sensor` specific support in `Entity` after this change
   - Reasonably simple to implement: It should be possible to simply rename all sensors' implementations of `state` and `unit_of_measurement` without any manual changes

In both cases:
 - Backwards compatibility for custom components is achieved by falling back to the old `Entity._async_write_ha_state` during a transition period.
 - Integrations can override the default behavior

Other considerations:
`climate`, `water_heater`, `weather` also do unit conversion, but it's already done in a clean way in the base components

**Comment by MartinHjelmare** [comment]
What will be the criteria for falling back to the old behavior for custom integrations?

**Reply by emontnemery** [reply]
We can keep the old conversion unchanged, and print a warning if it's executed.
If it's desirable to allow a sensor to report temperature in a custom unit, we could do the conversion only if the entity doesn't have `sensor_unit_of_measurement`.

Edit:
In addition to `sensor`, `air_quality` and `geo_location` also use `unit_of_measurement`. It's also the case for a handful of other entities.

**Comment by MartinHjelmare** [comment, marked as answer]
I'm for alternative number 2.

**Reply by emontnemery** [reply]
PR implementing alternative 2: https://github.com/home-assistant/core/pull/48261
